### DISCUSSION TITLE
DCE Demo Discussions: `custom-element` by Sasha Firsov

### DISCUSSION THREAD

**Discussion by Westbrook** [w3c/webcomponents-cg#86] [Ideas]
As presented on 18 March to the WCCG, `custom-element` is a proposed approach to declarative custom elements by @sashafirsov.

Links to docs TK as available.

Further discussion and questions below.

**Comment by sashafirsov** [comment]
[Live demo link](https://unpkg.com/@epa-wg/custom-element@0.0.17/index.html)

The primary differentiation of `custom-element` from other POCs on conceptual level:
* security focus: NO-JS app injection.  
* fully funstional semantics: template, browser APIs(storage, routing, networking), reactive ( events, state management )

It is not a DX or UX solution, rather closing the gap in Cyber Security on the web standard level.

**Reply by matthewp** [reply]
I love it.

**Comment by JRJurman** [comment]
On the surface this feels very different from existing Web Components, I would be curious to know / see the mapping from the APIs exposed here and those APIs in Web Components.

It would be also great to know what novel new stuff exists here, that isn't possible in class-based implementations, and what we imagine those implementations would look like if you wanted to build the class-based version.

**Reply by sashafirsov** [reply]
@JRJurman, The class-based implementation implied JS, which is a 
* not declarative syntax, hence should not be in the scope of **Declarative** Custom Element. Perhaps **inline declaration** is proper definition.
* It is not a streaming parser friendly, i.e. antipattern in HTML treatment.
* as JS is involved, it is not cross-vendor DCE code safe.
* I do not see the ability to do the initialization during the parsing by class-based implementation. In DCE it is a natural. 
* SSR, dynamic (re-)hydration , state sharing with JS in equation would require server side browser engine. With DCE that is not needed.

**Comment by sashafirsov** [comment]
@JRJurman 

> mapping from the APIs exposed here and those APIs in Web Components.
There is no mapping. Lifecycle is different and spread between tiers. The "API" as JS API is not applicable either. There is no "methods" or "properties". 

`custom-element` does use the WC convention internally. But exposing itself as one. Like the class, `observedAttributes`, events, etc. But it is not a part of DCE concept to be changed. Developers can extend DCE via WC registry though. And add any JS code to this class.
